[volume-6] 외부 시스템 장애 및 지연 대응

yeonsu00 opened the pull request:
## 📌 Summary
- PG 연동
- Resilience4j 적용
	- Circuit Breaker
	  - 슬라이딩 윈도우 20, 최소 호출 10, 실패율 60% 초과 시 OPEN 상태로 전환
	  - OPEN 상태에서 30초 대기 후 자동으로 HALF_OPEN 상태로 전환
	  - HALF_OPEN 상태에서 최대 5번의 호출만 허용하여 시스템 복구 여부 확인
	- Retry: 최대 2번 재시도 (총 3번 시도), 1초 간격
	- TimeLimiter: 3초 타임아웃 설정
	- 외부 시스템 장애 대응
		- Circuit Breaker Fallback: PG 시스템 장애 시 즉시 차단하고 Fallback 메서드 호출
		- Retry Fallback: 재시도 실패 시 PENDING 상태 응답 반환
	- 콜백 처리 및 상태 동기화
		- 콜백 수신: PG에서 전송하는 결제 완료/실패 콜백을 `PaymentV1Controller.handlePaymentCallback`에서 처리
		- 상태 업데이트: 콜백 수신 시 Order와 Payment 상태를 동기화

---

## 💬 Review Points

### 1. PG 결제 완료 후 콜백 전송 실패 시나리오 처리

PG 사에서 결제가 완료되었지만, 우리 서비스로 콜백을 보내는 과정에서 네트워크 오류가 발생하여 PG는 결제 완료 상태이지만 우리 서버는 결제 완료 사실을 모르는 상황일 때

1. 실무에서는 이런 상황을 어떻게 해결하시나요? 
   - 주기적으로 PG의 결제 상태 확인 API를 호출하여 동기화하는 스케줄러를 운영하시나요?

2. PG 상태 확인 API 호출 시 주의사항이 있나요?
   - 너무 빈번한 호출로 인한 Rate Limit, 동일 transactionKey에 대한 중복 처리 방지 등

---

### 2. 주문 만료 처리 및 스케줄러 운영

사용자가 주문을 생성했지만 일정 시간 내에 결제를 완료하지 않은 경우 주문 만료 처리를 어떻게 하시나요? 
`Order.createdAt`이 일정 시간 지난 주문들을 `OrderStatus.CANCELED`로 변경하는 스케줄러를 운영하시나요?

---

## ✅ Checklist

- [x] 불필요한 코드 제거
  
### **⚡ PG 연동 대응**

- [x]  PG 연동 API는 RestTemplate 혹은 FeignClient 로 외부 시스템을 호출한다.
- [x]  응답 지연에 대해 타임아웃을 설정하고, 실패 시 적절한 예외 처리 로직을 구현한다.
- [x]  결제 요청에 대한 실패 응답에 대해 적절한 시스템 연동을 진행한다.
- [x]  콜백 방식 + **결제 상태 확인 API**를 활용해 적절하게 시스템과 결제정보를 연동한다.

### **🛡 Resilience 설계**

- [x]  서킷 브레이커 혹은 재시도 정책을 적용하여 장애 확산을 방지한다.
- [x]  외부 시스템 장애 시에도 내부 시스템은 **정상적으로 응답**하도록 보호한다.
- [x]  콜백이 오지 않더라도, 일정 주기 혹은 수동 API 호출로 상태를 복구할 수 있다.
- [x]  PG 에 대한 요청이 타임아웃에 의해 실패되더라도 해당 결제건에 대한 정보를 확인하여 정상적으로 시스템에 반영한다.



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

## 릴리스 노트

* **New Features**
  * 주문 결제 기능이 추가되었습니다. 각 주문은 고유한 주문 키를 생성받으며, 외부 결제 시스템과의 연동을 통해 결제 처리가 가능합니다.
  * 결제 콜백 처리 기능으로 결제 성공/실패 시 주문 상태가 자동 업데이트되고, 실패 시 재고 및 포인트가 복구됩니다.
  * 결제 시스템의 안정성을 위해 회로 차단기, 타임아웃, 재시도 메커니즘이 적용되었습니다.

* **Chores**
  * 외부 결제 게이트웨이 시뮬레이터가 새로운 애플리케이션 모듈로 추가되었습니다.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

yeonsu00 committed:
feat: pg-simulator 모듈 추가 및 설정 추가

yeonsu00 committed:
Merge pull request #22 from yeonsu00/feature/week6-pg-simulator

pg-simulator 모듈 추가 및 설정 추가

yeonsu00 committed:
Merge branch 'main' into round-6

yeonsu00 committed:
feat: 결제 관련 도메인 추가

yeonsu00 committed:
feat: resilience4j, feign 설정 추가

yeonsu00 committed:
feat: 주문 요청 시 주문키 생성, Payment 저장 로직 추가

yeonsu00 committed:
feat: OrderInfo에 orderKey 필드 추가 및 from 메서드 오버로딩

yeonsu00 committed:
feat: 중복되는 resilience4j 의존성 제거

yeonsu00 committed:
feat: 주문 키 필드 추가 및 주문 상태 관리 메서드 구현

yeonsu00 committed:
feat: 결제 요청, 콜백 기능 추가

yeonsu00 committed:
feat: Kotlin 모듈 명시적 등록 및 PaymentRequest에 JSON 속성 추가

yeonsu00 committed:
feat: PaymentV1ApiSpec 추가

yeonsu00 committed:
refactor: PaymentCallbackRequest를 PaymentV1Dto로 이동

yeonsu00 committed:
feat: pg 결제 요청 재시도 후 fallback 시 주문, 포인트, 재고 원복 로직 추가

yeonsu00 committed:
feat: Payment에 orderId 제거, 관련 로직 수정

yeonsu00 committed:
test: PgPaymentClient Resilience4j 동작 테스트 추가

yeonsu00 committed:
Merge pull request #25 from yeonsu00/feature/week6-payments

Feature/week6 payments

yeonsu00 committed:
Merge pull request #26 from yeonsu00/round-6

[volume-6] 장애 및 지연 대응

looper88888 commented:
⚡ PG 연동 대응
 PG 연동 API는 RestTemplate 혹은 FeignClient 로 외부 시스템을 호출한다.
 응답 지연에 대해 타임아웃을 설정하고, 실패 시 적절한 예외 처리 로직을 구현한다.
 결제 요청에 대한 실패 응답에 대해 적절한 시스템 연동을 진행한다.
 콜백 방식 + 결제 상태 확인 API를 활용해 적절하게 시스템과 결제정보를 연동한다.
🛡 Resilience 설계
 서킷 브레이커 혹은 재시도 정책을 적용하여 장애 확산을 방지한다.
 외부 시스템 장애 시에도 내부 시스템은 정상적으로 응답하도록 보호한다.
 콜백이 오지 않더라도, 일정 주기 혹은 수동 API 호출로 상태를 복구할 수 있다.
 PG 에 대한 요청이 타임아웃에 의해 실패되더라도 해당 결제건에 대한 정보를 확인하여 정상적으로 시스템에 반영한다.
 
@coderabbitai 위 체크리스트를 확인해줘.

yeonsu00 committed:
refactor: pg 시스템 연동 패키지명 gateway로 변경

yeonsu00 committed:
refactor: PaymentCallbackRequest import를 PaymentV1Dto로 교체

yeonsu00 committed:
refactor: PaymentFacade에서 오케스트레이션 역할에 집중하도록 리팩토링

yeonsu00 committed:
refactor: OrderFacade에서 오케스트레이션 역할에 집중하도록 리팩토링, 결제 수단에 따른 로직 수정
- 포인트 결제 : 포인트 차감
- 카드 결제 : Payment 생성

yeonsu00 committed:
fix: pg 결제 시 포인트 관련 로직 삭제 및 카드 정보 업데이트

yeonsu00 committed:
refactor: 결제 요청 및 콜백 로직 수정
- PaymentService와 PaymentClient의 책임을 명확히 분리
- Payment 상태 수정은 콜백 로직이 아닌, 초기 결제 요청 후 처리

yeonsu00 committed:
refactor: PaymentService 카드 정보 처리 및 결제 요청 로직 개선

yeonsu00 committed:
refactor: PgPaymentClient 결제 요청 로직 개선 및 오류 처리 강화

yeonsu00 committed:
refactor: PgPaymentClient 및 PaymentClient 리팩토링, 오류 처리 로직 개선

yeonsu00 committed:
test: PgPaymentClientResilienceTest 재시도 불가능한 예외 처리 테스트 로직 수정

yeonsu00 committed:
feat: PG 결제 상태 동기화 스케줄러 및 콜백 미수신 복구 기능 추가

yeonsu00 committed:
Merge pull request #27 from yeonsu00/refactor/week6

Refactor/week6

yeonsu00 committed:
Merge pull request #28 from yeonsu00/round-6

Round 6

yeonsu00 commented:
⚡ PG 연동 대응
 PG 연동 API는 RestTemplate 혹은 FeignClient 로 외부 시스템을 호출한다.
 응답 지연에 대해 타임아웃을 설정하고, 실패 시 적절한 예외 처리 로직을 구현한다.
 결제 요청에 대한 실패 응답에 대해 적절한 시스템 연동을 진행한다.
 콜백 방식 + 결제 상태 확인 API를 활용해 적절하게 시스템과 결제정보를 연동한다.
🛡 Resilience 설계
 서킷 브레이커 혹은 재시도 정책을 적용하여 장애 확산을 방지한다.
 외부 시스템 장애 시에도 내부 시스템은 정상적으로 응답하도록 보호한다.
 콜백이 오지 않더라도, 일정 주기 혹은 수동 API 호출로 상태를 복구할 수 있다.
 PG 에 대한 요청이 타임아웃에 의해 실패되더라도 해당 결제건에 대한 정보를 확인하여 정상적으로 시스템에 반영한다.
 
@coderabbitai 위 체크리스트를 확인해줘.
